geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+            case "__version__" =>
+              new EncodeComplexAttribute {
+                def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) = e.writeInt(1)
+              }
+
+            case "__fid__" =>
+              new EncodeComplexAttribute {
+                def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) =
+                  e.writeString(Option(ca.getIdentifier) flatMap (i => Option(i.getID)) map (_.toString) getOrElse "")
+              }
+
+            case "simpleContent" => st match {
+              case n@Nullable(Primitive(t)) =>
+                val binding = bindEncPrimitive(AvroSchemas.nearestSimpleType(ct), t)
+                new EncodeComplexAttribute {
+                  def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) =
+                    Option(ca.getProperty(GEOT_SIMPLE_CONTENT)) match {
+                      case Some(value) if value.getValue != null =>
+                        n.writeNonNull(e)
+                        binding(e, value)
+                      case _ =>
+                        n.writeNull(e)
+                    }
+                }
+
+              case _ =>
+                throw new IllegalStateException(s"The property $name must be a nullable primitive type")
+            }
+
+            case other => st match {
+              case n: Nullable =>
+                new EncodeComplexAttribute {
+                  def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) = n.writeNull(e)
+                }
+              case _ => ???
+            }
+          }
+        } else {
+          descByName.get(name) match {
+            case Some(desc) =>
+              val binding = bindEncMany(st, desc.asInstanceOf[AttributeDescriptor])
+              new EncodeComplexAttribute {
+                def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) =
+                  binding(p.computeIfAbsent(name, arrayFactory), e)
+              }
+            case None =>
+              st match {
+                case n: Nullable =>
+                  new EncodeComplexAttribute {
+                    def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) = n.writeNull(e)
+                  }
+                case _ =>
+                  throw new IllegalStateException()
+              }
+          }
+        }
+      }.toArray
+
+
+      val ret = new EncodeProperty {
+        def apply(e: Encoder, p: Property) = {
+          val ca = p.asInstanceOf[ComplexAttribute]
+          val props = new JHashMap[Name, JArrayList[Property]]()
+          val pIt = ca.getProperties.iterator()
+          while (pIt.hasNext) {
+            val p = pIt.next()
+            props.computeIfAbsent(p.getName, arrayFactory).add(p)
+          }
+          val sz = bindings.length
+          var i = 0

```

Comment:
similar to above...perhaps a zipWithIndex if its not called often for scala-style iteratation